File: android/app/src/main/java/com/justspent/app/utils/LocalizationManager.kt
package com.justspent.app.utils

import android.content.Context
import org.json.JSONObject
import java.io.IOException

/**
 * LocalizationManager
 *
 * Loads localization strings from shared localizations.json
 * Single source of truth for cross-platform consistency
 */
class LocalizationManager private constructor(context: Context) {

    private val localizations: JSONObject
    private val platform = "android"

    init {
        localizations = loadLocalizations(context)
    }

    companion object {
        @Volatile
        private var instance: LocalizationManager? = null

        fun getInstance(context: Context): LocalizationManager {
            return instance ?: synchronized(this) {
                instance ?: LocalizationManager(context.applicationContext).also { instance = it }
            }
        }
    }

    // MARK: - Loading

    private fun loadLocalizations(context: Context): JSONObject {
        return try {
            val json = context.assets.open("localizations.json").bufferedReader().use { it.readText() }
            val jsonObject = JSONObject(json)
            println("✅ Loaded localizations.json version: ${jsonObject.optString("version", "unknown")}")
            jsonObject
        } catch (e: IOException) {
            println("❌ Failed to load localizations.json: ${e.message}")
            JSONObject()
        }
    }

    // MARK: - String Access

    /**
     * Get localized string by dot-notation path
     * Example: get("app.title") returns "Just Spent"
     */
    fun get(key: String): String {
        val components = key.split(".")
        var current: Any? = localizations

        for (component in components) {
            current = when (current) {
                is JSONObject -> current.opt(component)
                else -> null
            }
            if (current == null) {
                return "[$key]" // Return key in brackets if not found
            }
        }

        // Handle platform-specific strings
        if (current is JSONObject) {
            val platformValue = current.optString(platform)
            if (platformValue.isNotEmpty()) {
                return platformValue
            }
        }

        // Handle regular strings
        if (current is String) {
            return current
        }

        return "[$key]" // Return key in brackets if not found
    }

    // MARK: - Convenience Accessors

    val appTitle: String get() = get("app.title")
    val appSubtitle: String get() = get("app.subtitle")
    val appTotalLabel: String get() = get("app.totalLabel")

    val emptyStateNoExpenses: String get() = get("emptyState.noExpenses")
    val emptyStateTapVoiceButton: String get() = get("emptyState.tapVoiceButton")

    val buttonOK: String get() = get("buttons.ok")
    val buttonCancel: String get() = get("buttons.cancel")
    val buttonRetry: String get() = get("buttons.retry")

    val voiceListening: String get() = get("voice.listening")
    val voiceProcessing: String get() = get("voice.processing")

    val categoryFoodDining: String get() = get("categories.foodDining")
    val categoryGrocery: String get() = get("categories.grocery")
    val categoryTransportation: String get() = get("categories.transportation")
    val categoryShopping: String get() = get("categories.shopping")
    val categoryEntertainment: String get() = get("categories.entertainment")
    val categoryBills: String get() = get("categories.bills")
    val categoryHealthcare: String get() = get("categories.healthcare")
    val categoryEducation: String get() = get("categories.education")
    val categoryOther: String get() = get("categories.other")
    val categoryUnknown: String get() = get("categories.unknown")
}

File: ios/JustSpent/JustSpent/Common/LocalizationManager.swift
//
//  LocalizationManager.swift
//  JustSpent
//
//  Loads localization strings from shared localizations.json
//  Single source of truth for cross-platform consistency
//

import Foundation

/// Manages localized strings loaded from shared JSON file
class LocalizationManager {

    static let shared = LocalizationManager()

    private var localizations: [String: Any] = [:]
    private let platform = "ios"

    private init() {
        loadLocalizations()
    }

    // MARK: - Loading

    private func loadLocalizations() {
        let fileManager = FileManager.default
        var jsonURL: URL?

        // Try multiple paths to find the shared localization file

        // 1. Try from current working directory (works in tests)
        let currentDirPath = fileManager.currentDirectoryPath
        let sharedFromCurrent = "\(currentDirPath)/shared/localizations.json"
        if fileManager.fileExists(atPath: sharedFromCurrent) {
            jsonURL = URL(fileURLWithPath: sharedFromCurrent)
            print("📍 Found localizations.json in shared folder (from current dir)")
        }

        // 2. Try navigating up from bundle path (works in simulator)
        if jsonURL == nil, let bundlePath = Bundle.main.bundlePath as NSString? {
            // Navigate up from .app bundle to project root
            let projectRoot = bundlePath.deletingLastPathComponent
                .deletingLastPathComponent
                .deletingLastPathComponent
                .deletingLastPathComponent
            let sharedPath = "\(projectRoot)/shared/localizations.json"
            if fileManager.fileExists(atPath: sharedPath) {
                jsonURL = URL(fileURLWithPath: sharedPath)
                print("📍 Found localizations.json in shared folder (from bundle path)")
            }
        }

        // 3. Fallback: load from bundle (production build with file added to Xcode)
        if jsonURL == nil {
            jsonURL = Bundle.main.url(forResource: "localizations", withExtension: "json")
            if jsonURL != nil {
                print("📍 Found localizations.json in app bundle")
            }
        }

        guard let url = jsonURL,
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            print("❌ Failed to load localizations.json from any location")
            print("   Searched: current dir, bundle path, and app bundle")
            return
        }

        localizations = json
        print("✅ Loaded localizations.json version: \(json["version"] as? String ?? "unknown")")
    }

    // MARK: - String Access

    /// Get localized string by dot-notation path
    /// Example: get("app.title") returns "Just Spent"
    func get(_ key: String) -> String {
        let components = key.split(separator: ".").map(String.init)
        var current: Any? = localizations

        for component in components {
            guard let dict = current as? [String: Any] else {
                return "[\(key)]" // Return key in brackets if not found
            }
            current = dict[component]
        }

        // Handle platform-specific strings
        if let platformDict = current as? [String: String],
           let platformValue = platformDict[platform] {
            return platformValue
        }

        // Handle regular strings
        if let stringValue = current as? String {
            return stringValue
        }

        return "[\(key)]" // Return key in brackets if not found
    }

    // MARK: - Convenience Accessors

    var appTitle: String { get("app.title") }
    var appSubtitle: String { get("app.subtitle") }
    var appTotalLabel: String { get("app.totalLabel") }

    var emptyStateNoExpenses: String { get("emptyState.noExpenses") }
    var emptyStateTapVoiceButton: String { get("emptyState.tapVoiceButton") }

    var buttonOK: String { get("buttons.ok") }
    var buttonCancel: String { get("buttons.cancel") }
    var buttonRetry: String { get("buttons.retry") }

    var voiceListening: String { get("voice.listening") }
    var voiceProcessing: String { get("voice.processing") }

    var categoryFoodDining: String { get("categories.foodDining") }
    var categoryGrocery: String { get("categories.grocery") }
    var categoryTransportation: String { get("categories.transportation") }
    var categoryShopping: String { get("categories.shopping") }
    var categoryEntertainment: String { get("categories.entertainment") }
    var categoryBills: String { get("categories.bills") }
    var categoryHealthcare: String { get("categories.healthcare") }
    var categoryEducation: String { get("categories.education") }
    var categoryOther: String { get("categories.other") }
    var categoryUnknown: String { get("categories.unknown") }
}
